### PR TITLE
V3:  retry HTTP requests when there are transient failures

### PR DESCRIPTION
**Pull request by dtivel**
Resolve https://github.com/NuGet/Engineering/issues/295.

Also fixing:
* a bug where the underlying exception was obscured by a JSON parsing error ([example](https://github.com/NuGet/Engineering/issues/295#issuecomment-288201320)).
* a bug where `HttpResponseMessage` was not disposed in a timely manner

### CODE REVIEW

**Comment by loic-sharma** [comment]
Can we use ASP.NET Core's [HttpClientFactory](https://github.com/aspnet/HttpClientFactory) instead? This implementation has many more policies including "Retry, Circuit Breaker, Timeout, Bulkhead Isolation, and Fallback" (see [the documentation](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/http-requests?view=aspnetcore-2.1#use-polly-based-handlers)).

File: src/Catalog/RetryWithExponentialBackoff.cs [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+                ++_retries;
+
+                if (_retries < 31)
+                {
+                    _power = _power << 1;
+                }
+
+                int delay = (int)Math.Min(_delay.TotalMilliseconds * (_power - 1) / 2, _maximumDelay.TotalMilliseconds);
+
+                return Task.Delay(delay);
```
Comment (loic-sharma) [inline src/Catalog/RetryWithExponentialBackoff.cs]: Nit: maybe we should pass the `cancellationToken` to this method and not sleep if it's been triggered?

File: src/Catalog/CollectorHttpClient.cs [hunk 2]
```diff
@@ -86,25 +90,34 @@ public virtual Task<IGraph> GetGraphAsync(Uri address, bool readOnly, Cancellati
...
         {
-            var task = GetAsync(address, token);
+            var task = GetWithRetryAsync(address, token);
+
             return task.ContinueWith((t) =>
             {
                 try
                 {
-                    return task.Result.Content.ReadAsStringAsync().Result;
+                    using (var httpResponse = task.Result)
```
[the remark this answers is not in the document]
Comment (dtivel) [reply src/Catalog/CollectorHttpClient.cs]: I don't know why.  It's a vestige of the original implementation.

File: src/Catalog/RetryWithExponentialBackoff.cs [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+                {
+                    httpResponse = await client.GetAsync(address, cancellationToken);
+
+                    httpResponse.EnsureSuccessStatusCode();
+
+                    return httpResponse;
+                }
+                catch (HttpRequestException) when (IsTransientError(httpResponse))
+                {
+                    await backoff.Delay();
```
[the remark this answers is not in the document]
Comment (dtivel) [reply src/Catalog/RetryWithExponentialBackoff.cs]: Good catch.

**Comment by dtivel** [comment]
@loic-sharma, we'd need to update this project (and dependent projects) to at least .NET Framework 4.6.1, since the package you reference requires .NET Standard 2.0 or later.  Some dependencies are also not Authenticode signed.

File: src/Catalog/CollectorHttpClient.cs [hunk 4]
```diff
@@ -86,25 +87,24 @@ public virtual Task<IGraph> GetGraphAsync(Uri address, bool readOnly, Cancellati
...
         {
-            var task = GetAsync(address, token);
-            return task.ContinueWith((t) =>
+            try
             {
-                try
+                using (var httpResponse = await _retryStrategy.SendAsync(this, address, token))
                 {
-                    return task.Result.Content.ReadAsStringAsync().Result;
+                    return httpResponse.Content.ReadAsStringAsync().Result;
```
Comment (loic-sharma) [inline src/Catalog/CollectorHttpClient.cs]: Can we use an `await` here too?

**Pull request merged**